File: cpp/medium/#735AsteroidCollision.cpp
class Solution {
public:
    vector<int> asteroidCollision(vector<int>& asteroids) {
        vector<int> stack;

        for(int i = 0; i < asteroids.size(); i ++) {
            
            while(true) {
                if (asteroids[i] < 0 && stack.size() > 0 && stack.back() > 0) {
                    if (-asteroids[i] > stack.back()) {
                        stack.pop_back();
                        continue;
                    } else if (-asteroids[i] == stack.back()) {
                        stack.pop_back();
                        break;
                    }
                    break;
                } else {
                    stack.push_back(asteroids[i]);
                    break;
                }
            }
        }

        return stack;
    }
};

File: python/medium/#735AsteroidCollision.py
#TODO: revisit in C++
class Solution(object):
    def asteroidCollision(self, asteroids):
        """
        :type asteroids: List[int]
        :rtype: List[int]
        """
        stack = []

        for asteroid in asteroids:
            
            while stack and asteroid < 0 and stack[-1] > 0:
                leadingAsteroid = asteroid
                leftAsteroid = stack[-1]

                if abs(leftAsteroid) - abs(leadingAsteroid) < 0 and leadingAsteroid < 0:
                    winningAsteroid = leadingAsteroid
                    stack.pop()
                    continue
                if abs(leftAsteroid) - abs(leadingAsteroid) == 0:
                    # both asteroids annihilated 
                    stack.pop()
                break

            else:
                stack.append(asteroid)

        return stack